scylla-cdc-replicator/src/replicator_consumer.rs

```diff
@@ -0,0 +1,237 @@
+        let keys_iter = ReplicatorConsumer::get_keys_iter(&table_schema).await;
+        // Collect names of columns that are not clustering or partition key.
+        let other_columns = table_schema
+            .columns
+            .iter()
+            .filter(|column| {
+                column.1.kind != ColumnKind::Clustering && column.1.kind != ColumnKind::PartitionKey
+            })
+            .map(|column| column.0.clone())
+            .collect::<Vec<String>>();
+
+        // Clone, because the iterator is consumed.
+        let names = keys_iter.clone().join(",");
+        let markers = keys_iter
+            .map(|col_name| {
+                ReplicatorConsumer::bind_marker_for_type(
+                    &table_schema
+                        .columns
+                        .get(col_name)
+                        .expect("Column name not found")
+                        .type_,
+                )
+            })
+            .join(",");
+
+        let insert_query = dest_session
+            .prepare(format!(
+                "INSERT INTO {}.{} ({}) VALUES ({}) USING TTL ?",
+                dest_keyspace_name, dest_table_name, names, markers
+            ))
+            .await
+            .expect("Preparing insert query failed.");
+
+        ReplicatorConsumer {
+            dest_session,
+            dest_keyspace_name,
+            dest_table_name,
+            table_schema,
+            other_columns,
+            insert_query,
+        }
+    }
+
+    async fn get_keys_iter(
+        table_schema: &Table,
+    ) -> std::iter::Chain<Iter<'_, String>, Iter<'_, String>> {
+        table_schema
+            .partition_key
+            .iter()
+            .chain(table_schema.clustering_key.iter())
+    }
+
+    fn bind_marker_for_type(val_type: &CqlType) -> String {
+        if let CqlType::Tuple(vec) = val_type {
+            format!(
+                "({})",
+                vec.iter()
+                    .map(|val| ReplicatorConsumer::bind_marker_for_type(val))
+                    .join(",")
+            )
+        } else {
+            "?".to_string()
+        }
+    }
+
+    async fn update(&self, data: CDCRow<'_>) {
+        self.update_or_insert(data, false).await;
+    }
+
+    async fn insert(&self, data: CDCRow<'_>) {
+        self.update_or_insert(data, true).await;
+    }
+
+    async fn update_or_insert(&self, data: CDCRow<'_>, is_insert: bool) {
+        // [TODO]: Does not support deletions yet.
+        let keys_iter = ReplicatorConsumer::get_keys_iter(&self.table_schema).await;
+        let ttl = data.ttl.unwrap_or(0); // If data is inserted without TTL, setting it to 0 deletes existing TTL.
+        let timestamp = 0; // [TODO]: Map uuid to i64
+                           // [FIXME]: CqlValue does not have implemented FromIterator.
+        let mut values = keys_iter
+            .map(|col_name| data.get_value(col_name).unwrap())
+            .collect::<Vec<CqlValue>>();
```

Comment:
Small update: due to tiredness, I thought that the problem with `CqlValue` is more complex than it really is. Actually, `Value` trait is implemented for `CqlValue` so our only problem here is that we can't use iterators to map the values. This can be solved by implementing `FromIterator` trait for `CqlValue`. Alternatively, we could just not use the iterators and clone the values by using simple loop, but I think that we need the `Copy` trait, which is AFAIK also not implemented for `CqlValue` (I can't see it [in the documentation](https://docs.rs/scylla/latest/scylla/frame/response/result/enum.CqlValue.html#trait-implementations)). So, my question is: should we add one (or both) of these traits to `CqlValue` and if yes, should it be done in this crate or in the driver?
Bonus question, after we solve this issue: given a `key_values: Vec<CqlValue>`, is there a smart way to create a vector like this one: `[ttl, val] ++ key_values` (where `++` stands for vector concatenation) that neither copies `key_values' nor moves it? The query for value deletion will require different values than an update, so I'm not sure if we can afford to keep two separate vectors, or should merge the vectors as appropriate.

Reply:
> Bonus question, after we solve this issue: given a `key_values: Vec<CqlValue>`, is there a smart way to create a vector like this one: `[ttl, val] ++ key_values` (where `++` stands for vector concatenation) that neither copies `key_values' nor moves it?

Unfortunately, no. Perhaps if `Session::execute` or `Session::query` accepted an iterator then it would be possible (iterators can be chained together)

> The query for value deletion will require different values than an update, so I'm not sure if we can afford to keep two separate vectors, or should merge the vectors as appropriate.

Maybe you can just have two separate code paths which construct the values vector depending on the case?

Tip: if you want to efficiently construct a vector like `[ttl, val] ++ key_values` where `key_values` is an iterator, you can use the `extend` method:

```rust
// You can preallocate with `.reserve()` if you know how many values are in `key_values`
let mut values = vec![ttl, val];
values.extend(key_values);
```